src/acl/Asn.cc

```diff
@@ -410,48 +390,34 @@ asnAddNet(char *as_string, int as_number)
 
     debugs(53, 3, "asnAddNet: called for " << addr << "/" << mask );
 
-    rtentry_t *e = (rtentry_t *)xcalloc(1, sizeof(rtentry_t));
-
+    auto *e = new rtentry_t;
     e->e_addr.addr = addr;
-
     e->e_mask.addr = mask;
 
     rn = squid_rn_lookup(&e->e_addr, &e->e_mask, AS_tree_head);
 
     if (rn != NULL) {
-        asinfo = ((rtentry_t *) rn)->e_info;
+        auto &asinfo = reinterpret_cast<rtentry_t *>(rn)->e_info;
 
-        if (asinfo->as_number->find(as_number)) {
+        if (std::find(asinfo.begin(), asinfo.end(), as_number) != asinfo.end()) {
             debugs(53, 3, "asnAddNet: Ignoring repeated network '" << addr << "/" << bitl << "' for AS " << as_number);
         } else {
             debugs(53, 3, "asnAddNet: Warning: Found a network with multiple AS numbers!");
-
-            for (Tail = &asinfo->as_number; *Tail; Tail = &(*Tail)->next);
-            q = new CbDataList<int> (as_number);
-
-            *(Tail) = q;
-
-            e->e_info = asinfo;
+            asinfo.emplace_back(as_number);
         }
+        delete e;
+
     } else {
-        q = new CbDataList<int> (as_number);
-        asinfo = (as_info *)xmalloc(sizeof(as_info));
-        asinfo->as_number = q;
         squid_rn_addroute(&e->e_addr, &e->e_mask, AS_tree_head, e->e_nodes);
-        rn = squid_rn_match(&e->e_addr, AS_tree_head);
-        assert(rn != NULL);
-        e->e_info = asinfo;
-    }
-
-    if (rn == 0) {      /* assert might expand to nothing */
-        xfree(asinfo);
-        delete q;
-        xfree(e);
-        debugs(53, 3, "asnAddNet: Could not add entry.");
-        return 0;
+        if ((rn = squid_rn_match(&e->e_addr, AS_tree_head))) {
+            e->e_info.emplace_back(as_number);
```

Comment:
AFAICT, this `e->e_info.emplace_back()` statement is the only case where the allocated `e` is not explicitly deleted in the PR version of asnAddNet(). Let's focus on this specific case. I cannot understand where `e` (i.e. its allocated memory) goes and what code will eventually free it. Knowing the answers is important for several reasons, but primarily because this PR _changes_ how rtentry_t needs to be deleted (and copied/allocated/etc.). Can you answer those two questions, please?

Reply:
As far as I am able to understand the radix code squid_rn_addroute() tries to absorb "e" into the AS_tree_head tree structure by setting the e->e_nodes member values.

All other cases "e" has remains owned by this function.

Reply:
> squid_rn_addroute() tries to absorb "e" into the AS_tree_head tree structure

AFAICT, squid_rn_addroute() is unaware of the actual `e` type so `e` absorption cannot work correctly for non-trivial rtentry_t objects. This PR makes rtentry_t non-trivial.

Reply:
Please note that I do _not_ know whether your [understanding](https://github.com/squid-cache/squid/pull/720#discussion_r482646904) regarding squid_rn_addroute() is correct. I am only saying that _if_ it is correct, then this PR is wrong.

Reply:
If you are worried about entries not being delete()'d after they are absorbed into the tree there is the destroyRadixNode() walker which extracts entries and calls delete on them.

Reply:
> If you are worried about entries not being delete()'d after they are absorbed into the tree 

That is one of many things I am worried about. Absorption means taking full control. One cannot have full control of a C++ object without knowing its type unless that type is trivial. This PR makes rtentry_t non-trivial.

The existing/official radix API is badly broken, but we can assume that it more-or-less works or is largely unused because there is no stream of bug reports about the related functionality. Quality replacements of that API should be welcomed. Surgical bug fixes correctly using the existing API should be welcomed. This PR is currently neither.

> there is the destroyRadixNode() walker which extracts entries and calls delete on them.

destroyRadixNode() does not call `delete` on rtentry_t objects we are discussing here, but even if it were doing that, my concerns would remain because we have no guarantees that all these non-trivial objects were properly adsorbed and handled in the low-level radix code that separates our shiny type-specific `new` and `delete`.

Rough summary: Given an `absorb(void*)` API, we cannot change the type of the object (which address we pass to absorb()) from trivial to non-trivial (without changing the API). In most cases, we cannot change the type of the object at all -- even changing one trivial type to another trivial type would require changing the code behind absorb()! Absorbing usually requires understanding the object structure. AFAICT, that is the case with squid_rn_addroute(). Without compiler-checked type safety (which we do not have here due to forced casts), we cannot alter the types of objects passed through that absorbing API, especially in non-trivial ways.